lib/pkcs11/portable/espressif/esp32_devkitc_esp_wrover_kit/aws_pkcs11_pal.c

```diff
@@ -51,73 +53,60 @@ enum eObjectHandles
     eAwsDeviceCertificate,
     eAwsCodeSigningKey
 };
-
 /*-----------------------------------------------------------*/
 
-static void initialize_spiffs()
+static void initialize_nvs_partition()
 {
-    static bool spiffs_inited = false;
-
-    ESP_LOGI( TAG, "Initializing SPIFFS" );
+    static bool nvs_inited;
 
     portENTER_CRITICAL();
-    if ( spiffs_inited == true )
-    {
+    if (nvs_inited == true) {
         portEXIT_CRITICAL();
         return;
     }
 
-    esp_vfs_spiffs_conf_t conf =
-    {
-        .base_path = "/spiffs",
-        .partition_label = pkcsFLASH_PARTITION,
-        .max_files = 5,
-        .format_if_mount_failed = true
-    };
+    ESP_LOGI(TAG, "Initializing NVS partition: \"%s\"", NVS_PART_NAME);
 
-    /* Use settings defined above to initialize and mount SPIFFS filesystem. */
-    /* Note: esp_vfs_spiffs_register is an all-in-one convenience function. */
-    esp_err_t ret = esp_vfs_spiffs_register( &conf );
 
-    if ( ret == ESP_OK )
-    {
-        spiffs_inited = true;
-    }
-    portEXIT_CRITICAL();
 
-    if( ret != ESP_OK )
-    {
-        if( ret == ESP_FAIL )
-        {
-            ESP_LOGE( TAG, "Failed to mount or format filesystem" );
+#if CONFIG_NVS_ENCRYPTION
+    if (esp_flash_encryption_enabled()) {
+        const esp_partition_t *key_part = esp_partition_find_first(
+                ESP_PARTITION_TYPE_DATA, ESP_PARTITION_SUBTYPE_DATA_NVS_KEYS, NULL);
+        assert(key_part && "NVS key partition not found");
+
+        nvs_sec_cfg_t cfg;
+        esp_err_t err = nvs_flash_read_security_cfg(key_part, &cfg);
+        if (err == ESP_ERR_NVS_KEYS_NOT_INITIALIZED) {
+            ESP_LOGI(TAG, "NVS key partition empty, generating keys");
+            nvs_flash_generate_keys(key_part, &cfg);
+        } else {
+            ESP_ERROR_CHECK(err);
         }
-        else if( ret == ESP_ERR_NOT_FOUND )
-        {
-            ESP_LOGE( TAG, "Failed to find SPIFFS partition" );
+
+        esp_err_t ret = nvs_flash_secure_init_partition(NVS_PART_NAME, &cfg);
+        if (ret == ESP_ERR_NVS_NO_FREE_PAGES || ret == ESP_ERR_NVS_NEW_VERSION_FOUND) {
+            ESP_ERROR_CHECK(nvs_flash_erase_partition(NVS_PART_NAME));
+            ret = nvs_flash_secure_init_partition(NVS_PART_NAME, &cfg);
         }
-        else
-        {
-            ESP_LOGE( TAG, "Failed to initialize SPIFFS (%d)", ret );
+        ESP_ERROR_CHECK(ret);
+    } else {
+#endif // CONFIG_NVS_ENCRYPTION
+        esp_err_t ret = nvs_flash_init_partition(NVS_PART_NAME);
+        if (ret == ESP_ERR_NVS_NO_FREE_PAGES || ret == ESP_ERR_NVS_NEW_VERSION_FOUND) {
```

Comment:
What causes a new version of the NVS to be found?   Is erasing the flash partition a desirable outcome for this?

Reply:
If someone is using NVS partition contents from older release (IDF v3.0 and below), and NVS component with `multiblob feature` support enabled (supports writing key value greater than 1984 bytes with `CONFIG_MP_BLOB_SUPPORT`), then it may result in version incompatibility. So, in this case erasing NVS partition is only option for further functionality to work.

Interim, `sdkconfig.defaults` should include `CONFIG_MP_BLOB_SUPPORT=y` as available in original PR, certain certificates can be bigger than 1984 bytes to be written to NVS storage partition.